wren-ui/src/apollo/server/backgrounds/chart.ts

```diff
@@ -0,0 +1,218 @@
+import { ChartStatus } from '@server/models/adaptor';
+import { IWrenAIAdaptor } from '@server/adaptors/wrenAIAdaptor';
+import {
+  IThreadResponseRepository,
+  ThreadResponse,
+} from '@server/repositories';
+import { getLogger } from '@server/utils/logger';
+
+const logger = getLogger('ChartBackgroundTracker');
+logger.level = 'debug';
+
+const isFinalized = (status: ChartStatus) => {
+  return (
+    status === ChartStatus.FINISHED ||
+    status === ChartStatus.FAILED ||
+    status === ChartStatus.STOPPED
+  );
+};
+
+export class ChartBackgroundTracker {
+  private tasks: Record<number, ThreadResponse> = {};
+  private intervalTime: number;
+  private wrenAIAdaptor: IWrenAIAdaptor;
+  private threadResponseRepository: IThreadResponseRepository;
+  private runningJobs = new Set();
+
+  constructor({
+    wrenAIAdaptor,
+    threadResponseRepository,
+  }: {
+    wrenAIAdaptor: IWrenAIAdaptor;
+    threadResponseRepository: IThreadResponseRepository;
+  }) {
+    this.wrenAIAdaptor = wrenAIAdaptor;
+    this.threadResponseRepository = threadResponseRepository;
+    this.intervalTime = 1000;
+    this.start();
+  }
+
+  private start() {
+    logger.info('Chart background tracker started');
+    setInterval(() => {
+      const jobs = Object.values(this.tasks).map(
+        (threadResponse) => async () => {
+          // check if same job is running
+          if (this.runningJobs.has(threadResponse.id)) {
+            return;
+          }
+
+          // mark the job as running
+          this.runningJobs.add(threadResponse.id);
+
+          // get the chart detail
+          const chartDetail = threadResponse.chartDetail;
+
+          // get the latest result from AI service
+          const result = await this.wrenAIAdaptor.getChartResult(
+            chartDetail.queryId,
+          );
+
+          // check if status change
+          if (chartDetail.status === result.status) {
+            // mark the job as finished
+            logger.debug(
+              `Job ${threadResponse.id} chart status not changed, finished`,
+            );
+            this.runningJobs.delete(threadResponse.id);
+            return;
+          }
+
+          // update database
+          const updatedChartDetail = {
+            queryId: chartDetail.queryId,
```

Comment:
why do we need to update the queryId? will it change?

Reply:
I remove it in 126b0845

Reply:
@onlyjackfrost I added the `queryId` back, because it will cause the `queryId` disappear because the column is `jsonb`